README.md
### A `FormData` polyfill for the browser ...and a module for NodeJS (`New!`)

```bash
npm install formdata-polyfill
```

The browser polyfill will likely have done its part already, and i hope you stop supporting old browsers c",)<br>
But NodeJS still laks a proper FormData<br>The good old form-data package is a very old and isn't spec compatible and dose some abnormal stuff to construct and read FormData instances that other http libraries are not happy about when it comes to follow the spec.

### The NodeJS / ESM version
- The modular (~2.3 KiB minified uncompressed) version of this package is independent of any browser stuff and don't patch anything
- It's as pure/spec compatible as it possible gets the test are run by WPT.
- It's compatible with [node-fetch](https://github.com/node-fetch/node-fetch).
- It have higher platform dependencies as it uses classes, symbols, ESM & private fields
- Only dependency it has is [fetch-blob](https://github.com/node-fetch/fetch-blob)

```js
// Node example
import fetch from 'node-fetch'
import File from 'fetch-blob/file.js'
import { fileFromSync } from 'fetch-blob/from.js'
import { FormData } from 'formdata-polyfill/esm.min.js'

const file = fileFromSync('./README.md')
const fd = new FormData()

fd.append('file-upload', new File(['abc'], 'hello-world.txt'))
fd.append('file-upload', file)

// it's also possible to append file/blob look-a-like items
// if you have streams coming from other destinations
fd.append('file-upload', {
  size: 123,
  type: '',
  name: 'cat.mp4',
  stream() { return stream },
  [Symbol.toStringTag]: 'File'
})

fetch('https://httpbin.org/post', { method: 'POST', body: fd })
```

----

It also comes with way to convert FormData into Blobs - it's not something that every developer should have to deal with.
It's mainly for [node-fetch](https://github.com/node-fetch/node-fetch) and other http library to ease the process of serializing a FormData into a blob and just wish to deal with Blobs instead (Both Deno and Undici adapted a version of this [formDataToBlob](https://github.com/jimmywarting/FormData/blob/5ddea9e0de2fc5e246ab1b2f9d404dee0c319c02/formdata-to-blob.js) to the core and passes all WPT tests run by the browser itself)
```js
import { Readable } from 'node:stream'
import { FormData, formDataToBlob } from 'formdata-polyfill/esm.min.js'

const blob = formDataToBlob(new FormData())
fetch('https://httpbin.org/post', { method: 'POST', body: blob })

// node built in http and other similar http library have to do:
const stream = Readable.from(blob.stream())
const req = http.request('http://httpbin.org/post', {
  method: 'post',
  headers: {
    'Content-Length': blob.size,
    'Content-Type': blob.type
  }
})
stream.pipe(req)
```

PS: blob & file that are appended to the FormData will not be read until any of the serialized blob read-methods gets called
...so uploading very large files is no biggie

### Browser polyfill

usage:

```js
import 'formdata-polyfill' // that's it
```

The browser polyfill conditionally replaces the native implementation rather than fixing the missing functions,
since otherwise there is no way to get or delete existing values in the FormData object.
Therefore this also patches `XMLHttpRequest.prototype.send` and `fetch` to send the `FormData` as a blob,
and `navigator.sendBeacon` to send native `FormData`.

I was unable to patch the Response/Request constructor
so if you are constructing them with FormData then you need to call `fd._blob()` manually.

```js
new Request(url, {
  method: 'post',
  body: fd._blob ? fd._blob() : fd
})
```

Dependencies
---

If you need to support IE <= 9 then I recommend you to include eligrey's [blob.js]
(which i hope you don't - since IE is now dead)

<details>
    <summary>Updating from 2.x to 3.x</summary>

Previously you had to import the polyfill and use that,
since it didn't replace the global (existing) FormData implementation.
But now it transparently calls `_blob()` for you when you are sending something with fetch or XHR,
by way of monkey-patching the `XMLHttpRequest.prototype.send` and `fetch` functions.

So you maybe had something like this:

```javascript
var FormData = require('formdata-polyfill')
var fd = new FormData(form)
xhr.send(fd._blob())
```

There is no longer anything exported from the module
(though you of course still need to import it to install the polyfill),
so you can now use the FormData object as normal:

```javascript
require('formdata-polyfill')
var fd = new FormData(form)
xhr.send(fd)
```

</details>



Native Browser compatibility (as of 2021-05-08)
---
Based on this you can decide for yourself if you need this polyfill.

[![screenshot](https://user-images.githubusercontent.com/1148376/117550329-0993aa80-b040-11eb-976c-14e31f1a3ba4.png)](https://developer.mozilla.org/en-US/docs/Web/API/FormData#Browser_compatibility)



This normalizes support for the FormData API:

 - `append` with filename
 - `delete()`, `get()`, `getAll()`, `has()`, `set()`
 - `entries()`, `keys()`, `values()`, and support for `for...of`
 - Available in web workers (just include the polyfill)

  [npm-image]: https://img.shields.io/npm/v/formdata-polyfill.svg
  [npm-url]: https://www.npmjs.com/package/formdata-polyfill
  [blob.js]: https://github.com/eligrey/Blob.js


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

### Added
- Start writing a changelog
- ensure right number of arguments are provided
- Started running test against WPTs own test
### Changed
- Updated fetch-blob dep
- downgraded optional chaining to support node 12.20
### Removed
- removed travis build process file
- removed empty npmignore file
### Fixed
- Pass blobs property bag over to file when converting it


## [4.0.5] - 2021-06-19

[Unreleased]: https://github.com/jimmywarting/FormData/compare/4.0.5...HEAD
[4.0.5]: https://github.com/jimmywarting/FormData/compare/4.0.4...4.0.5


test/test.js
/* global it describe assert expect Blob FormData */

const native = window.FormData
window.FormData = undefined

class Polyfill {
  constructor (args) {
    return new FormData(args)
  }
}

const NativeFile = window.File
// Want to simulate the kind of error we get in
// old safari & ie when constructing a file
window.File = new Proxy(NativeFile, {
  construct (target, args) {
    throw new Error()
  }
})

;[native, Polyfill].forEach(FormData => {
  function createFormData (...args) {
    const fd = new FormData()
    for (const arg of args) {
      fd.append(...arg)
    }
    return fd
  }

  function createForm (str) {
    var form = document.createElement('form')
    form.innerHTML = str
    return new FormData(form)
  }

  // Some test are imported from
  // http://bit.ly/2zyG5yZ

  describe('FormData ' + (FormData === native ? 'native' : 'polyfill'), () => {
    describe('append', () => {
      it('throwOnFewArgs', () => {
        expect(() => {
          var fd = new FormData()
          fd.append('key')
        }).throw()
      })

      // #78
      it('testEmptyFileInput', () => {
        const fd = createForm(`<input name=foo type=file>`)
        assert.equal(fd.has('foo'), true)
        assert.equal(fd.get('foo').type, 'application/octet-stream')
      })

      it('testFormDataAppend1', () => {
        assert.equal(createFormData(['key', 'value1']).get('key'), 'value1')
      })

      it('testFormDataAppend2', () => {
        assert.equal(createFormData(['key', 'value2'], ['key', 'value1']).get('key'), 'value2')
      })

      it('testFormDataAppendUndefined1', () => {
        assert.equal(createFormData(['key', undefined]).get('key'), 'undefined')
      })

      it('testFormDataAppendUndefined2', () => {
        assert.equal(createFormData(['key', undefined], ['key', 'value1']).get('key'), 'undefined')
      })

      it('testFormDataAppendNull1', () => {
        assert.equal(createFormData(['key', null]).get('key'), 'null')
      })

      it('testFormDataAppendNull2', () => {
        assert.equal(createFormData(['key', null], ['key', 'value1']).get('key'), 'null')
      })

      it('testFormDataAppendToForm1', () => {
        var fd = new FormData(document.createElement('form'))
        fd.append('key', 'value1')
        assert.equal(fd.get('key'), 'value1')
      })

      it('testFormDataAppendToForm2', () => {
        var fd = new FormData(document.createElement('form'))
        fd.append('key', 'value2')
        fd.append('key', 'value1')
        assert.equal(fd.get('key'), 'value2')
      })

      it('testFormDataAppendToFormUndefined1', () => {
        var fd = new FormData(document.createElement('form'))
        fd.append('key', undefined)
        assert.equal(fd.get('key'), 'undefined')
      })

      it('testFormDataAppendToFormUndefined2', () => {
        var fd = new FormData(document.createElement('form'))
        fd.append('key', undefined)
        fd.append('key', 'value1')
        assert.equal(fd.get('key'), 'undefined')
      })

      it('testFormDataAppendToFormNull1', () => {
        var fd = new FormData(document.createElement('form'))
        fd.append('key', null)
        assert.equal(fd.get('key'), 'null')
      })

      it('testFormDataAppendToFormNull2', () => {
        var fd = new FormData(document.createElement('form'))
        fd.append('key', null)
        fd.append('key', 'value1')
        assert.equal(fd.get('key'), 'null')
      })

      it('testFormDataAppendEmptyBlob', () => {
        var fd = createFormData(['key', new Blob(), 'blank.txt']).get('key')
        assert.equal(fd.name, 'blank.txt')
        assert.equal(fd.type, '')
        assert.equal(fd.size, 0)
      })
    })

    describe('has', () => {
      it('throwOnFewArgs', () => {
        expect(() => {
          var fd = new FormData()
          fd.has()
        }).throw()
      })

      it('FormData.has()', () => {
        var fd = new FormData()
        fd.append('n1', 'value')
        assert.equal(fd.has('n1'), true)
        assert.equal(fd.has('n2'), false)
        fd.append('n2', 'value')
        assert.equal(fd.has('n1'), true)
        assert.equal(fd.has('n2'), true)
        fd.append('n3', new Blob(['content']))
        assert.equal(fd.has('n3'), true)
      })
    })

    describe('entries', () => {
      it('should return the keys/values/entres as they are appended', () => {
        const fd = createFormData(
          ['keyA', 'val1'],
          ['keyA', 'val2'],
          ['keyB', 'val3'],
          ['keyA', 'val4']
        )

        assert.deepEqual([...fd.keys()], ['keyA', 'keyA', 'keyB', 'keyA'])
        assert.deepEqual([...fd.values()], ['val1', 'val2', 'val3', 'val4'])
        assert.deepEqual([...fd], [
          ['keyA', 'val1'],
          ['keyA', 'val2'],
          ['keyB', 'val3'],
          ['keyA', 'val4']
        ])
      })
    })

    describe('set', () => {
      it('Overwrite first matching key', () => {
        const fd = createFormData(
          ['keyA', 'val1'],
          ['keyA', 'val2'],
          ['keyB', 'val3'],
          ['keyA', 'val4']
        )
        fd.set('keyA', 'val3')
        assert.deepEqual([...fd], [['keyA', 'val3'], ['keyB', 'val3']])
      })

      it('appends value when no matching', () => {
        const fd = createFormData(
          ['keyB', 'val3']
        )
        fd.set('keyA', 'val3')
        assert.deepEqual([...fd], [['keyB', 'val3'], ['keyA', 'val3']])
      })
    })

    describe('delete', () => {
      it('throwOnFewArgs', () => {
        expect(() => {
          var fd = new FormData()
          fd.delete()
        }).throw()
      })

      it('FormData.delete()', () => {
        var fd = new FormData()
        fd.append('name', 'value')
        assert.equal(fd.has('name'), true)
        fd.delete('name')
        assert.equal(fd.has('name'), false)

        fd.append('name', new Blob(['content']))
        assert.equal(fd.has('name'), true)
        fd.delete('name')
        assert.equal(fd.has('name'), false)

        fd.append('n1', 'v1')
        fd.append('n2', 'v2')
        fd.append('n1', 'v3')
        fd.delete('n1')
        assert.equal(fd.has('n1'), false)

        assert.deepEqual([...fd], [['n2', 'v2']])
      })
    })

    describe('filename', () => {
      // #43
      // Old ie don't have Symbol.toStringTag and the polyfill was
      // therefore not able to change the
      // `Object.prototype.toString.call` to return correct type of the polyfilled
      // File constructor
      it('Shold return correct filename with File', () => {
        const fd = createFormData(['key', new NativeFile([], 'doc.txt')])
        const mockFile = fd.get('key')
        assert.equal('doc.txt', mockFile.name)
      })

      it('Shold return correct filename with Blob filename', () => {
        const fd = createFormData(['key', new Blob(), 'doc.txt'])
        const mockFile = fd.get('key')
        assert.equal('doc.txt', mockFile.name)
      })

      it('Shold return correct filename with just Blob', () => {
        const fd = createFormData(['key', new Blob()])
        const mockFile = fd.get('key')
        assert.equal('blob', mockFile.name)
      })
    })

    describe('linebreaks', () => {
      const form = document.createElement('form')
      const textarea = document.createElement('textarea')
      textarea.name = 'key'

      // Native FormData normalizes linefeeds in textareas to CRLF
      // In order to be consistent the polyfill should do the same
      it('Should convert LF to CRLF for textareas', () => {
        textarea.value = '\n'
        form.appendChild(textarea)
        const fd = new FormData(form)
        const value = fd.get('key')
        assert.equal('\r\n', value)
      })

      it('Should convert CR to CRLF for textareas', () => {
        textarea.value = '\r'
        form.appendChild(textarea)
        const fd = new FormData(form)
        const value = fd.get('key')
        assert.equal('\r\n', value)
      })

      it('Should normalize mixed linefeeds to CRLF for textareas', () => {
        textarea.value = 'a\n\ra\r\na\n\r\n\r\n\r\n\na\r\r'
        form.appendChild(textarea)
        const fd = new FormData(form)
        const value = fd.get('key')
        assert.equal('a\r\n\r\na\r\na\r\n\r\n\r\n\r\n\r\na\r\n\r\n', value)
      })

      it('Should not convert LF to CRLF when provided by append', () => {
        const fd = createFormData(['key', '\n'])
        const value = fd.get('key')
        assert.equal('\n', value)
      })

      it('Should not convert CR to CRLF when provided by append', () => {
        const fd = createFormData(['key', '\r'])
        const value = fd.get('key')
        assert.equal('\r', value)
      })
    })

    describe('disabled', () => {
      it('Shold not include disabled inputs', () => {
        const fd = createForm(
          `<input disabled name=foo value=bar>`
          )
          assert.deepEqual([...fd], [])
        })

      // #110
      it('Shold not include disabled fieldset inputs', () => {
        const fd = createForm(
          `<fieldset disabled><input name=foo value=bar></fieldset>`
        )
        assert.deepEqual([...fd], [])
      })

      // #56
      it('Select elements where the option is both selected and disabled should not be included', () => {
        const fd = createForm(`
          <select multiple name="example">
            <option selected disabled value="foo">Please choose one</option>
          </select>
        `)

        assert.equal([...fd].length, 0)
      })
    })

    describe('constructor', () => {
      // #45
      it('Shold return selected items', () => {
        const fd = createForm(`
          <select multiple name="example">
            <option selected value="volvo">Volvo</option>
            <option selected value="saab">Saab</option>
            <option value="mercedes">Mercedes</option>
            <option value="audi">Audi</option>
          </select>
        `)

        assert.deepEqual([...fd], [['example', 'volvo'], ['example', 'saab']])
      })

      // #62
      it('Should not add buttons to FormData', () => {
        const fd = createForm(`
          <input type="text" name="username" value="bob">
          <input type="submit" value="rename" name="action">
          <input type="submit" value="find_n_delete" name="action">
        `)

        assert.deepEqual([...fd], [['username', 'bob']])
      })
    })

    describe.skip('XHR', () => {
      // #86
      it('Shouldn not add Content-Type header if already present', next => {
        const formData = createFormData(['key', 'value1'])
        const xhr = new XMLHttpRequest()
        xhr.open('POST', 'https://httpbin.org/post')
        xhr.setRequestHeader('Content-Type', 'text/plain')
        xhr.responseType = 'json'
        xhr.onload = () => {
          assert.equal(xhr.response.headers['Content-Type'], 'text/plain')
          next()
        }
        xhr.send(formData)
      })
    })
  })
})


test/http-loader.js
// https-loader.mjs
import { get } from 'https';

export function resolve(specifier, context, defaultResolve) {
  const { parentURL = null } = context;

  // Normally Node.js would error on specifiers starting with 'https://', so
  // this hook intercepts them and converts them into absolute URLs to be
  // passed along to the later hooks below.
  if (specifier.startsWith('https://')) {
    return {
      url: specifier
    };
  } else if (parentURL && parentURL.startsWith('https://')) {
    return {
      url: new URL(specifier, parentURL).href
    };
  }

  // Let Node.js handle all other specifiers.
  return defaultResolve(specifier, context, defaultResolve);
}

export function getFormat(url, context, defaultGetFormat) {
  // This loader assumes all network-provided JavaScript is ES module code.
  if (url.startsWith('https://')) {
    return {
      format: 'module'
    };
  }

  // Let Node.js handle all other URLs.
  return defaultGetFormat(url, context, defaultGetFormat);
}

export function getSource(url, context, defaultGetSource) {
  // For JavaScript to be loaded over the network, we need to fetch and
  // return it.
  if (url.startsWith('https://')) {
    const {searchParams} = new URL(url)

    return new Promise((resolve, reject) => {
      let data = 'globalThis.self = globalThis;'
      get(url, async res => {
        for await (const chunk of res) data += chunk;
        searchParams.getAll('expose').forEach(variable => {
          data += `\n;globalThis.${variable} = ${variable};`
        })
        resolve({ source: data });
      }).on('error', (err) => reject(err));
    });
  }

  // Let Node.js handle all other URLs.
  return defaultGetSource(url, context, defaultGetSource);
}


formdata-to-blob.js
/*! formdata-polyfill. MIT License. Jimmy Wärting <https://jimmy.warting.se/opensource> */

const escape = str => str.replace(/\n/g, '%0A').replace(/\r/g, '%0D').replace(/"/g, '%22')

/**
 * pure function to convert any formData instance to a Blob
 * instances synchronous without reading all of the files
 *
 * @param {FormData|*} formData an instance of a formData Class
 * @param {Blob|*} [BlobClass=Blob] the Blob class to use when constructing it
 */
export function formDataToBlob (formData, BlobClass = Blob) {
  const boundary = ('----formdata-polyfill-' + Math.random())
  const chunks = []
  const prefix = `--${boundary}\r\nContent-Disposition: form-data; name="`

  for (let [name, value] of formData) {
    if (typeof value === 'string') {
      chunks.push(prefix + escape(name) + `"\r\n\r\n${value.replace(/\r(?!\n)|(?<!\r)\n/g, '\r\n')}\r\n`)
    } else {
      chunks.push(
        prefix + escape(name) + `"; filename="${escape(value.name)}"\r\n` +
        `Content-Type: ${value.type || 'application/octet-stream'}\r\n\r\n`,
        value,
        '\r\n'
      )
    }
  }

  chunks.push(`--${boundary}--`)

  return new BlobClass(chunks, {
    type: 'multipart/form-data; boundary=' + boundary
  })
}


FormData.js
/* formdata-polyfill. MIT License. Jimmy Wärting <https://jimmy.warting.se/opensource> */

/* global FormData self Blob File */
/* eslint-disable no-inner-declarations */

if (typeof Blob !== 'undefined' && (typeof FormData === 'undefined' || !FormData.prototype.keys)) {
  const global = typeof globalThis === 'object'
    ? globalThis
    : typeof window === 'object'
      ? window
      : typeof self === 'object' ? self : this

  // keep a reference to native implementation
  const _FormData = global.FormData

  // To be monkey patched
  const _send = global.XMLHttpRequest && global.XMLHttpRequest.prototype.send
  const _fetch = global.Request && global.fetch
  const _sendBeacon = global.navigator && global.navigator.sendBeacon
  // Might be a worker thread...
  const _match = global.Element && global.Element.prototype

  // Unable to patch Request/Response constructor correctly #109
  // only way is to use ES6 class extend
  // https://github.com/babel/babel/issues/1966

  const stringTag = global.Symbol && Symbol.toStringTag

  // Add missing stringTags to blob and files
  if (stringTag) {
    if (!Blob.prototype[stringTag]) {
      Blob.prototype[stringTag] = 'Blob'
    }

    if ('File' in global && !File.prototype[stringTag]) {
      File.prototype[stringTag] = 'File'
    }
  }

  // Fix so you can construct your own File
  try {
    new File([], '') // eslint-disable-line
  } catch (a) {
    global.File = function File (b, d, c) {
      const blob = new Blob(b, c || {})
      const t = c && void 0 !== c.lastModified ? new Date(c.lastModified) : new Date()

      Object.defineProperties(blob, {
        name: {
          value: d
        },
        lastModified: {
          value: +t
        },
        toString: {
          value () {
            return '[object File]'
          }
        }
      })

      if (stringTag) {
        Object.defineProperty(blob, stringTag, {
          value: 'File'
        })
      }

      return blob
    }
  }

  function ensureArgs (args, expected) {
    if (args.length < expected) {
      throw new TypeError(`${expected} argument required, but only ${args.length} present.`)
    }
  }

  function normalizeArgs (name, value, filename) {
    if (value instanceof Blob) {
      filename = filename !== undefined
      ? String(filename + '')
      : typeof value.name === 'string'
      ? value.name
      : 'blob'

      if (value.name !== filename || Object.prototype.toString.call(value) === '[object Blob]') {
        value = new File([value], filename)
      }
      return [String(name), value]
    }
    return [String(name), String(value)]
  }

  // normalize line feeds for textarea
  // https://html.spec.whatwg.org/multipage/form-elements.html#textarea-line-break-normalisation-transformation
  function normalizeLinefeeds (value) {
    return value.replace(/\r?\n|\r/g, '\r\n')
  }

  function each (arr, cb) {
    for (let i = 0; i < arr.length; i++) {
      cb(arr[i])
    }
  }

  const escape = str => str.replace(/\n/g, '%0A').replace(/\r/g, '%0D').replace(/"/g, '%22')

  /**
   * @implements {Iterable}
   */
  class FormDataPolyfill {
    /**
     * FormData class
     *
     * @param {HTMLElement=} form
     */
    constructor (form) {
      this._data = []

      const self = this

      form && each(form.elements, elm => {
        if (
          !elm.name ||
          elm.disabled ||
          elm.type === 'submit' ||
          elm.type === 'button' ||
          elm.matches('form fieldset[disabled] *')
        ) return

        if (elm.type === 'file') {
          const files = elm.files && elm.files.length
            ? elm.files
            : [new File([], '', { type: 'application/octet-stream' })] // #78

          each(files, file => {
            self.append(elm.name, file)
          })
        } else if (elm.type === 'select-multiple' || elm.type === 'select-one') {
          each(elm.options, opt => {
            !opt.disabled && opt.selected && self.append(elm.name, opt.value)
          })
        } else if (elm.type === 'checkbox' || elm.type === 'radio') {
          if (elm.checked) self.append(elm.name, elm.value)
        } else {
          const value = elm.type === 'textarea' ? normalizeLinefeeds(elm.value) : elm.value
          self.append(elm.name, value)
        }
      })
    }

    /**
     * Append a field
     *
     * @param   {string}           name      field name
     * @param   {string|Blob|File} value     string / blob / file
     * @param   {string=}          filename  filename to use with blob
     * @return  {undefined}
     */
    append (name, value, filename) {
      ensureArgs(arguments, 2)
      this._data.push(normalizeArgs(name, value, filename))
    }

    /**
     * Delete all fields values given name
     *
     * @param   {string}  name  Field name
     * @return  {undefined}
     */
    delete (name) {
      ensureArgs(arguments, 1)
      const result = []
      name = String(name)

      each(this._data, entry => {
        entry[0] !== name && result.push(entry)
      })

      this._data = result
    }

    /**
     * Iterate over all fields as [name, value]
     *
     * @return {Iterator}
     */
    * entries () {
      for (var i = 0; i < this._data.length; i++) {
        yield this._data[i]
      }
    }

    /**
     * Iterate over all fields
     *
     * @param   {Function}  callback  Executed for each item with parameters (value, name, thisArg)
     * @param   {Object=}   thisArg   `this` context for callback function
     * @return  {undefined}
     */
    forEach (callback, thisArg) {
      ensureArgs(arguments, 1)
      for (const [name, value] of this) {
        callback.call(thisArg, value, name, this)
      }
    }

    /**
     * Return first field value given name
     * or null if non existent
     *
     * @param   {string}  name      Field name
     * @return  {string|File|null}  value Fields value
     */
    get (name) {
      ensureArgs(arguments, 1)
      const entries = this._data
      name = String(name)
      for (let i = 0; i < entries.length; i++) {
        if (entries[i][0] === name) {
          return entries[i][1]
        }
      }
      return null
    }

    /**
     * Return all fields values given name
     *
     * @param   {string}  name  Fields name
     * @return  {Array}         [{String|File}]
     */
    getAll (name) {
      ensureArgs(arguments, 1)
      const result = []
      name = String(name)
      each(this._data, data => {
        data[0] === name && result.push(data[1])
      })

      return result
    }

    /**
     * Check for field name existence
     *
     * @param   {string}   name  Field name
     * @return  {boolean}
     */
    has (name) {
      ensureArgs(arguments, 1)
      name = String(name)
      for (let i = 0; i < this._data.length; i++) {
        if (this._data[i][0] === name) {
          return true
        }
      }
      return false
    }

    /**
     * Iterate over all fields name
     *
     * @return {Iterator}
     */
    * keys () {
      for (const [name] of this) {
        yield name
      }
    }

    /**
     * Overwrite all values given name
     *
     * @param   {string}    name      Filed name
     * @param   {string}    value     Field value
     * @param   {string=}   filename  Filename (optional)
     * @return  {undefined}
     */
    set (name, value, filename) {
      ensureArgs(arguments, 2)
      name = String(name)
      const result = []
      const args = normalizeArgs(name, value, filename)
      let replace = true

      // - replace the first occurrence with same name
      // - discards the remaining with same name
      // - while keeping the same order items where added
      each(this._data, data => {
        data[0] === name
          ? replace && (replace = !result.push(args))
          : result.push(data)
      })

      replace && result.push(args)

      this._data = result
    }

    /**
     * Iterate over all fields
     *
     * @return {Iterator}
     */
    * values () {
      for (const [, value] of this) {
        yield value
      }
    }

    /**
     * Return a native (perhaps degraded) FormData with only a `append` method
     * Can throw if it's not supported
     *
     * @return {FormData}
     */
    ['_asNative'] () {
      const fd = new _FormData()

      for (const [name, value] of this) {
        fd.append(name, value)
      }

      return fd
    }

    /**
     * [_blob description]
     *
     * @return {Blob} [description]
     */
    ['_blob'] () {
        const boundary = '----formdata-polyfill-' + Math.random(),
          chunks = [],
          p = `--${boundary}\r\nContent-Disposition: form-data; name="`
        this.forEach((value, name) => typeof value == 'string'
          ? chunks.push(p + escape(normalizeLinefeeds(name)) + `"\r\n\r\n${normalizeLinefeeds(value)}\r\n`)
          : chunks.push(p + escape(normalizeLinefeeds(name)) + `"; filename="${escape(value.name)}"\r\nContent-Type: ${value.type||"application/octet-stream"}\r\n\r\n`, value, `\r\n`))
        chunks.push(`--${boundary}--`)
        return new Blob(chunks, {
          type: "multipart/form-data; boundary=" + boundary
        })
    }

    /**
     * The class itself is iterable
     * alias for formdata.entries()
     *
     * @return  {Iterator}
     */
    [Symbol.iterator] () {
      return this.entries()
    }

    /**
     * Create the default string description.
     *
     * @return  {string} [object FormData]
     */
    toString () {
      return '[object FormData]'
    }
  }

  if (_match && !_match.matches) {
    _match.matches =
      _match.matchesSelector ||
      _match.mozMatchesSelector ||
      _match.msMatchesSelector ||
      _match.oMatchesSelector ||
      _match.webkitMatchesSelector ||
      function (s) {
        var matches = (this.document || this.ownerDocument).querySelectorAll(s)
        var i = matches.length
        while (--i >= 0 && matches.item(i) !== this) {}
        return i > -1
      }
  }

  if (stringTag) {
    /**
     * Create the default string description.
     * It is accessed internally by the Object.prototype.toString().
     */
    FormDataPolyfill.prototype[stringTag] = 'FormData'
  }

  // Patch xhr's send method to call _blob transparently
  if (_send) {
    const setRequestHeader = global.XMLHttpRequest.prototype.setRequestHeader

    global.XMLHttpRequest.prototype.setRequestHeader = function (name, value) {
      setRequestHeader.call(this, name, value)
      if (name.toLowerCase() === 'content-type') this._hasContentType = true
    }

    global.XMLHttpRequest.prototype.send = function (data) {
      // need to patch send b/c old IE don't send blob's type (#44)
      if (data instanceof FormDataPolyfill) {
        const blob = data['_blob']()
        if (!this._hasContentType) this.setRequestHeader('Content-Type', blob.type)
        _send.call(this, blob)
      } else {
        _send.call(this, data)
      }
    }
  }

  // Patch fetch's function to call _blob transparently
  if (_fetch) {
    global.fetch = function (input, init) {
      if (init && init.body && init.body instanceof FormDataPolyfill) {
        init.body = init.body['_blob']()
      }

      return _fetch.call(this, input, init)
    }
  }

  // Patch navigator.sendBeacon to use native FormData
  if (_sendBeacon) {
    global.navigator.sendBeacon = function (url, data) {
      if (data instanceof FormDataPolyfill) {
        data = data['_asNative']()
      }
      return _sendBeacon.call(this, url, data)
    }
  }

  global['FormData'] = FormDataPolyfill
}
